src/west/manifest.py

```diff
@@ -430,69 +490,179 @@ def _load(self, manifest, path_hint):
 
-        return ManifestProject(path=slf.get('path', path_hint),
-                               topdir=self.topdir,
-                               west_commands=slf.get('west-commands'))
+        return mp
+
+    def _assert_imports_ok(self):
+        # Sanity check that we aren't calling code that does importing
+        # if the flags tell us not to.
+        #
+        # Could be deleted if this feature stabilizes and we never hit
+        # this assertion.
+
+        assert not self._import_flags & ImportFlag.IGNORE
+
+    def _import_from_self(self, mp, imp, projects):
+        # Recursive helper to import projects from the manifest repository.
+        #
+        # - mp: the ManifestProject
+        # - imp: "self: import: <imp>" value from manifest data
+        # - projects: ordered map of Project instances we've already got
+        #
+        # All data is read from the file system. Requests to read
+        # files which don't exist or aren't ordinary files/directories
+        # raise MalformedManifest.
+        #
+        # This is unlike importing from projects -- for projects, data
+        # are read from Git (treating it as a content-addressable file
+        # system) with a fallback on self._importer.
+
+        self._assert_imports_ok()
+
+        self.has_imports = True
+
+        imptype = type(imp)
+        if imptype == bool:
+            self._malformed('got "self: import: {}" of boolean'.format(imp))
+        elif imptype == str:
+            self._import_path_from_self(mp, imp, projects)
+        elif imptype == list:
+            for subimp in imp:
+                self._import_from_self(mp, subimp, projects)
+        elif imptype == dict:
+            self._import_map_from_self(mp, imp, projects)
+        else:
+            self._malformed('{}: "self: import: {}" has invalid type {}'.
+                            format(mp.abspath, imp, imptype))
+
+    def _import_path_from_self(self, mp, imp, projects):
+        if mp.abspath:
+            # Fast path, when we're working inside a fully initialized
+            # topdir.
+            log.dbg('manifest repository root:', mp.abspath,
+                    level=log.VERBOSE_EXTREME)
+            repo_root = Path(mp.abspath)
+        else:
+            # Fallback path, which is needed by at least west init. If
+            # this happens too often, something may be wrong with how
+            # we've implemented this. We'd like to avoid too many git
+            # commands, as subprocesses are slow on windows.
+            log.dbg('searching for the manifest repository root',
+                    level=log.VERBOSE_EXTREME)
+            repo_root = Path(mp.git('rev-parse --show-toplevel',
+                                    capture_stdout=True,
+                                    cwd=str(Path(self.path).parent)).
+                             stdout[:-1].      # chop off newline
+                             decode('utf-8'))  # hopefully this is safe
```

Comment:
I think if that path is found for most west commands, we're likely already in the fast path above, not the fallback path. Can you come up with a case where that's not true?

Reply:
> I think if that path is found for most west commands, we're likely already in the fast path above, not the fallback path. Can you come up with a case where that's not true?

No, I was just wondering that because you added a fallback mechanism here, you might have foreseen a risk, and in that case I just wondered if `.west/config` could be a better option instead of resolving to `git`

Especially as one can do a `west init -l <local-only-filesystem>`.
I don't think we mandate the manifest repo to be a git repo, it could be just local files.

Reply:
> No, I was just wondering that because you added a fallback mechanism here, you might have foreseen a risk, and in that case I just wondered if `.west/config` could be a better option instead of resolving to `git`

It wasn't foreseen, it was run into in practice in west init, actually. I have no crystal ball ;).

> Especially as one can do a `west init -l <local-only-filesystem>`.
> I don't think we mandate the manifest repo to be a git repo, it could be just local files.

That still assumes it's a manifest repository, i.e. a git repository.